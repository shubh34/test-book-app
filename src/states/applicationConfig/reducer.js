export const defaultState = {
  books: [],
  isApiCallInProgress: [],
  hasError: false
};

const reducer = (state = defaultState, action) => {
  return state;
};
export default reducer;
