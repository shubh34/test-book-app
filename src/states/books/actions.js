import { RSAA } from 'redux-api-middleware';
import endpoints from '../../configs/endpoints';

export const FETCH_BOOKS_REQUEST = 'GET_BOOKS_REQUEST';
export const FETCH_BOOKS_SUCCESS = 'GET_BOOKS_SUCCESS';
export const FETCH_BOOKS_FAILURE = 'GET_BOOKS_FAILURE';

export const SET_PAGE_NUMBER_TO_FETCH_BOOKS = 'SET_PAGE_NUMBER_TO_FETCH_BOOKS';
export const SET_PER_PAGE_BOOKS_TO_FETCH = 'SET_PER_PAGE_BOOKS_TO_FETCH';
export const SET_SEARCH_CRITERIA_TO_FETCH_BOOKS = 'SET_SEARCH_CRITERIA_TO_FETCH_BOOKS';

export const fetchBookRequest = (page) => ({
  type: FETCH_BOOKS_REQUEST,
  meta: {
    page
  }
});
export const fetchBookSuccess = (page) => ({
  type: FETCH_BOOKS_SUCCESS,
  meta: {
    page
  }
});
export const fetchBookFailure = (page) => ({
  type: FETCH_BOOKS_FAILURE,
  meta: {
    page
  }
});

export const fetchBooks = ({
  page = 1,
  itemsPerPage = 20,
  filters = [{ type: 'all', values: [''] }]
} = {}) => {
  const endPoint = endpoints.fetchBooks;
  const url = endPoint.resolve();
  return {
    [RSAA]: {
      endpoint: url,
      method: endPoint.method,
      body: JSON.stringify({
        page,
        itemsPerPage,
        filters
      }),
      types: [fetchBookRequest(page), fetchBookSuccess(page), fetchBookFailure(page)]
    }
  };
};

export const setPageNumberToFetchBook = (page) => ({
  type: SET_PAGE_NUMBER_TO_FETCH_BOOKS,
  page
});

export const setPerPageToFetchBooks = (itemsPerPage) => ({
  type: SET_PER_PAGE_BOOKS_TO_FETCH,
  itemsPerPage
});

export const setSearchCriteriaToFetchBook = (search) => ({
  type: SET_SEARCH_CRITERIA_TO_FETCH_BOOKS,
  filters: [{ type: 'all', values: [search] }]
});
