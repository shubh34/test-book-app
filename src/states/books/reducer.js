import { FETCH_BOOKS_REQUEST, FETCH_BOOKS_SUCCESS, FETCH_BOOKS_FAILURE } from './actions';

export const pageDefaultState = {
  lists: [],
  isApiCallInProgress: false,
  hasError: false,
  error: undefined
};
export const defaultState = {
  books: {},
  isApiCallInProgress: false,
  hasError: false,
  error: undefined
};

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case FETCH_BOOKS_REQUEST:
      return {
        ...state,
        books: {
          ...state.books,
          [action.meta.page]: {
            ...pageDefaultState,
            isApiCallInProgress: true,
            hasError: false,
            error: undefined
          }
        }
      };
    case FETCH_BOOKS_SUCCESS:
      return {
        ...state,
        books: {
          ...state.books,
          [action.meta.page]: {
            ...pageDefaultState,
            lists: action.payload.books
          }
        }
      };
    case FETCH_BOOKS_FAILURE:
      return {
        ...state,
        books: {
          ...state.books,
          [action.meta.page]: {
            ...pageDefaultState,
            hasError: true
          }
        }
      };
    default:
      return state;
  }
};
export default reducer;
