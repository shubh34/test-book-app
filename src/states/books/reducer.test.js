import reducer from './reducer';
import { FETCH_BOOKS_REQUEST, FETCH_BOOKS_SUCCESS, FETCH_BOOKS_FAILURE } from './actions';

describe('Reducer Books', () => {
  it('FETCH_BOOKS_REQUEST should update api in call for requested page', () => {
    const action = {
      type: FETCH_BOOKS_REQUEST,
      meta: { page: 1 }
    };
    const updateState = reducer(undefined, action);
    expect(updateState.books[1].isApiCallInProgress).toBe(true);
    expect(updateState).toMatchSnapshot();
  });
  it('FETCH_BOOKS_SUCCESS should update books list of requested page', () => {
    const action = {
      type: FETCH_BOOKS_SUCCESS,
      payload: { books: [{ id: 'book-id' }], count: 2 },
      meta: { page: 1 }
    };
    const updateState = reducer(undefined, action);
    expect(updateState).toMatchSnapshot();
  });
  it('FETCH_BOOKS_FAILURE should update error for requested page', () => {
    const action = {
      type: FETCH_BOOKS_FAILURE,
      meta: { page: 1 }
    };
    const updateState = reducer(undefined, action);
    expect(updateState.books[1].hasError).toBe(true);
    expect(updateState).toMatchSnapshot();
  });
});
