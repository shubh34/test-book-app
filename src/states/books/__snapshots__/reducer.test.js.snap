// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer Books FETCH_BOOKS_FAILURE should update error for requested page 1`] = `
Object {
  "books": Object {
    "1": Object {
      "error": undefined,
      "hasError": true,
      "isApiCallInProgress": false,
      "lists": Array [],
    },
  },
  "meta": Object {
    "count": 0,
    "filters": Array [],
    "itemsPerPage": 20,
    "page": 1,
  },
}
`;

exports[`Reducer Books FETCH_BOOKS_REQUEST should update api in call for requested page 1`] = `
Object {
  "books": Object {
    "1": Object {
      "error": undefined,
      "hasError": false,
      "isApiCallInProgress": true,
      "lists": Array [],
    },
  },
  "meta": Object {
    "count": 0,
    "filters": Array [],
    "itemsPerPage": 20,
    "page": 1,
  },
}
`;

exports[`Reducer Books FETCH_BOOKS_SUCCESS should update books list of requested page 1`] = `
Object {
  "books": Object {
    "1": Object {
      "error": undefined,
      "hasError": false,
      "isApiCallInProgress": false,
      "lists": Array [
        Object {
          "id": "book-id",
        },
      ],
    },
  },
  "meta": Object {
    "count": 2,
    "filters": Array [],
    "itemsPerPage": 20,
    "page": 1,
  },
}
`;
